django/core/servers/basehttp.py

```diff
@@ -92,6 +107,10 @@ def cleanup_headers(self):
         if self.headers.get('Connection') == 'close':
             self.request_handler.close_connection = True
 
+    def close(self):
+        super().close()
+        self.get_stdin()._read_limited()
```

Comment:
I don't suppose it actually makes a difference, but would we not do this in the other order? (i.e. the `super()` call at the end.)

Reply:
In super.close we call close method on different entities that still have access to request and request body I don't think we want to exhaust input stream before everyone who had access to it releases it. I can't come up with a scenario when something can use that request body in close method but for me this oder looks right because the ServerHandler limited stream should be a hidden abstraction to all underneath code.

Reply:
My query was prompted by general life-cycle considerations: in general, when creating an object we mostly call `super()` first then add our customizations; when tearing down, we remove our customizations then call `super()`.

At this point we want to make sure our request body is fully consumed before exiting... so I'd expect to do our work, call `_read_limited()`, and then call `super()`.

* You had it the other way round in b08ada36f30fef1df948911ca50ac28c58eb39a4.
* The superclass [`simple_server.ServerHandler`](https://github.com/python/cpython/blob/e6b247c8e524dbe5fc03b3492f628d0d5348bc49/Lib/wsgiref/simple_server.py#L28-L38) does similar, calling the parent implementation last.  

I was wondering why you changed it?

Reply:
> At this point we want to make sure our request body is fully consumed before exiting

No, in gunicorn they consume body of previous request on next request in this session to make sure that no extra work was done if connection is closed. So no one should expect that body is consumed by us at the moment when we call close. 

In fact it doesn't matter much until we have some closable object that read from stdin on closing. Should I swap the lines?

Reply:
> Should I swap the lines?

Well, let's get it right. 🙂 Sorry if you have to explain it to me... 

> ...in gunicorn they ...

But this is just for runserver right? Gunicorn is a whole different topic. They have they're own handling. No? 

AFAICS (here) there's no difference between a request where we actually `read()` the request body (to consume it) and one where we make sure it was read right at the end, so that it doesn't become part of the next request. 

I'm struggling to think of a case where it would be at all sensible to read from stdin when closing. As I said to begin, I don't suppose it actually makes a difference for that reason. But still curious as to the _why_ of the change?

Reply:
> Well, let's get it right. 🙂 Sorry if you have to explain it to me...

Ok, no problem

> But this is just for runserver right? Gunicorn is a whole different topic. They have they're own handling. No?

gunicorn was just an example that shows that the data can be read at latest moment not at any particular close event or something like that. 

> As I said to begin, I don't suppose it actually makes a difference for that reason. 

Yes that change is purely how it maps into my logic, I don't get the idea of clearing request data before we finished everything else what was working with that request but because it shouldn't break anything anyway I'm not against that change.

> But still curious as to the why of the change?

Previous version of that pr was pushed by @apollo13 I commented about that line swap on that commit but no one replied. So on refactoring I swapped it because for me it look obvious that the logic of clearing post data shouldn't leak to underneath code from ServerHandler, to me it looks like this: you wrap request into proxy object to clear it later -> you gives that object to underneath code to work on -> you make sure that everything is finished working on that proxy object (close call) -> you clear things up (_read_limited call).

Reply:
Hi @kalekseev. In the end I'd rather we went with the other ordering here. (I favoured it, Florian favoured, wsgiref handler favours it,... — **if** a case comes up that needs it the other way round we can address it then, with tests.) 

I've adjusted and rebased. Will merge later on, after the CI finishes. 

Thanks again for your super input! "First-time contributor" — Welcome aboard! 🎉